src/simularium/WebsocketClient.ts

```diff
@@ -0,0 +1,317 @@
+    private serverIp: string;
+    private serverPort: number;
+    public connectionTimeWaited: number;
+    public connectionRetries: number;
+    protected jsonMessageHandlers: Map<NetMessageEnum, (NetMessage) => void>;
+    protected binaryMessageHandlers: Map<
+        NetMessageEnum,
+        (MessageEvent) => void | ((MessageEventLike) => void)
+    >;
+    protected logger: ILogger;
+    public handleError: (error: FrontEndError) => void | (() => void);
+
+    public constructor(
+        opts?: NetConnectionParams,
+        errorHandler?: (error: FrontEndError) => void
+    ) {
+        this.webSocket = null;
+        this.jsonMessageHandlers = new Map<
+            NetMessageEnum,
+            (NetMessage) => void
+        >();
+        this.binaryMessageHandlers = new Map<
+            NetMessageEnum,
+            (NetMessage) => void
+        >();
+        this.serverIp = opts && opts.serverIp ? opts.serverIp : "localhost";
+        this.serverPort = opts && opts.serverPort ? opts.serverPort : 9002;
+        this.connectionTimeWaited = 0;
+        this.connectionRetries = 0;
+        this.handleError = errorHandler || (() => {});
+
+        this.logger = jsLogger.get("netconnection");
+        this.logger.setLevel(jsLogger.DEBUG);
+
+        // Frees the reserved backend in the event that the window closes w/o disconnecting
+        window.addEventListener("beforeunload", this.onClose.bind(this));
+    }
+
+    /**
+     * WebSocket State
+     */
+    private socketIsConnecting(): boolean {
+        return (
+            this.webSocket !== null &&
+            this.webSocket.readyState === this.webSocket.CONNECTING
+        );
+    }
+
+    public socketIsValid(): boolean {
+        return !(
+            this.webSocket === null ||
+            this.webSocket.readyState === this.webSocket.CLOSED
+        );
+    }
+
+    private socketIsConnected(): boolean {
+        return (
+            this.webSocket !== null &&
+            this.webSocket.readyState === this.webSocket.OPEN
+        );
+    }
+
+    /**
+     *   Websocket Message Handling
+     * */
+    public addBinaryMessageHandler(
+        messageType: NetMessageEnum,
+        handler: (msg: MessageEvent | MessageEventLike) => void
+    ): void {
+        this.binaryMessageHandlers[messageType.valueOf()] = handler;
+    }
+
+    public addJsonMessageHandler(
+        messageType: NetMessageEnum,
+        handler: (msg: NetMessage) => void
+    ): void {
+        this.jsonMessageHandlers[messageType] = handler;
+    }
+
+    protected onMessage(event: MessageEvent | MessageEventLike): void {
+        // where we receive websocket messages
+        if (!this.socketIsValid()) {
+            return;
+        }
+
+        if (event.data instanceof ArrayBuffer) {
+            // Handle binary message
+            const floatView = new Float32Array(event.data);
+            const binaryMsgType = floatView[0];
+            for (let msgType in this.binaryMessageHandlers) {
```

Comment:
instead of a for loop, I wonder if this could be written as something like
```
if this.binaryMessageHandlers.has(msgType)
  this.binaryMessageHandlers[msgType](event)
```